### DISCUSSION TITLE
Delays in user input

### DISCUSSION THREAD

**Discussion by Markkaz** [bevyengine/bevy#16335] [Q&A] [answered]
I'm working for a week now with the Bevy game engine, and having some trouble with input delays. Both with the UI as well as directly with keyboard input. And I was wondering what I was doing wrong. It seems that there is always a few frames delay.
However, when I try the examples on the Bevy website, there doesn't seem to be any delay input when I run the WASM versions of it. So my expectation is that it has something to do with my compiler settings.

So for now, I have the following source code:
```
impl Plugin for MenuPlugin {
    fn build(&self, app: &mut App) {
        app.configure_sets(Update, MenuUpdateSet.run_if(in_state(GameState::Menu)));
        app.add_systems(OnEnter(GameState::Menu), (
            init_menu
        ));
        app.add_systems(Update, (update_menu).in_set(MenuUpdateSet));
    }
}

fn init_menu(mut commands: Commands) {
    commands.spawn(NodeBundle {
        style: Style {
            width: Val::Percent(100.),
            height: Val::Percent(100.),
            justify_content: JustifyContent::Center,
            align_items: AlignItems::Center,
            flex_direction: FlexDirection::Column,
            ..default()
        },
        ..default()
    })
    .with_children(|parent| {
        add_button(parent, "Play");
        add_button(parent, "Exit");
    });
}

fn add_button(parent: &mut ChildBuilder, title: &str) {
    parent.spawn(ButtonBundle {
        style: Style {
            width: Val::Percent(150.),
            height: Val::Px(65.),
            justify_content: JustifyContent::Center,
            align_items: AlignItems::Center,
            ..default()
        },
        background_color: menu::NORMAL.into(),
        ..default()
    })
        .with_children(|parent| {
            parent.spawn(TextBundle::from_section(
                title,
                TextStyle {
                    font_size: 40.,
                    color: Color::srgb(0.9, 0.9, 0.9),
                    ..default()
                }
            ));
        });
}

fn update_menu(
    mut interaction_query: Query<
        (&Interaction, &mut BackgroundColor),
        (Changed<Interaction>, With<Button>),
    >
) {
    for (interaction, mut color) in &mut interaction_query {
        match *interaction {
            Interaction::Pressed => *color = menu::PRESSED.into(),
            Interaction::Hovered => *color = menu::HOVERED.into(),
            Interaction::None => *color = menu::NORMAL.into(),
        }
    }
}
```

My Cargo.toml file looks like this, for optimalisations:
```
[package]
name = "pong"
version = "0.1.0"
edition = "2021"

[profile.dev]
opt-level = 1

[profile.dev.package."*"]
opt-level = 3

[dependencies]
bevy = "0.14.2"
bevy_rapier2d = "0.27.0"
```

I seem to have the same delays both when I compile in dev mode as well as release mode.

**Comment by alice-i-cecile** [comment]
I don't think this is likely to be a problem in the compiler settings: if this problem is widespread it's almost certainly a bug in winit or our integration with winit.

**Reply by Markkaz** [reply]
As an experiment I compiled the same code towards WASM. And indeed, the WASM code doesn't have the delay in input. Proving the hypothesis that it has something to do with the Windows integration.

**Reply by Markkaz** [reply]
Just did another test on a Macbook Pro: MacOS Sequoia 15.1
Also here the experience is smooth, no problem with input delays.

**Comment by alice-i-cecile** [comment]
Can you say more about how you're measuring the input delay? That would be very helpful in general, but especially for others to check across different platforms.

**Reply by Markkaz** [reply]
I didn't do an official benchmark. Just build a small pong application, and the delay in the input is very noticable. There seems to be almost half a second between the input and the change actually happening on the screen.
The code I used for the test: https://github.com/Markkaz/Pong

My Windows PC is running Windows 11.

**Reply by alice-i-cecile** [reply, marked as answer]
Have you experimented with https://github.com/aevyrie/bevy_framepace? Poor framepacing may be the culprit here.

**Reply by Markkaz** [reply]
I wasn't aware of this library, but it actually fixed the problem!
Thank you for all the help, really appreciated!

**Reply by aevyrie** [reply]
@alice-i-cecile it might be worth upstreaming this plugin, even if it is naive. The latency and power savings benefits are pretty large for such a small amount of code.

**Reply by aevyrie** [reply]
It does have some issues, but the lag does not leave a great first impression either.

**Reply by alice-i-cecile** [reply]
Yep, this is on my radar, and very much something I'd like to upstream.

**Comment by AmorLudum** [comment]
I also encountered input delay problems on Windows.
My solution is:
1. set the wgpu backend as DX12
2. set the present mode of the window as AutoNoVsync
```rust
App::new()
    .add_plugins(DefaultPlugins.set(
        RenderPlugin {
            render_creation: RenderCreation::Automatic(
                WgpuSettings {
                    backends: Some(Backends::DX12),
                    ..default()
                }
            ),
            ..default()
        }
    ).set(
        WindowPlugin {
            primary_window: Some(
                Window {
                    present_mode: PresentMode::AutoNoVsync,
                    ..default()
                }
            ),
            ..default()
        }
    ))
```

**Comment by Markkaz** [comment]
I upgraded my project to Bevy version 0.15. Unfortunately the bevy_framespace plugin doesn't work in this version. 
I tried @AmorLudum 's solution, and that actually solves the problem as well. You only need to change the present_mode to AutoNoVSync. No need to set the WGPU backend.

```rust
App::new()
        .add_plugins(
            DefaultPlugins.set(
                WindowPlugin {
                    primary_window: Some(
                        Window {
                            present_mode: PresentMode::AutoNoVsync,
                            ..default()
                        }
                    ),
                    ..default()
                }
            )
        )
```

**Reply by guyguy2001** [reply]
This worked for me - thank you so much!

(Writing down "display lag" as a keyword so that future google searches will have an easier time finding this 🙏)

**Comment by dubrowgn** [comment]
Related: https://github.com/bevyengine/bevy/discussions/21807
